src/hotspot/share/opto/loopopts.cpp

```diff
@@ -4120,3 +4121,204 @@ bool PhaseIdealLoop::duplicate_loop_backedge(IdealLoopTree *loop, Node_List &old
+      if (current == last_ur) {
+        // Expect all uses to be outside the loop, except phi.
+        for (DUIterator_Fast kmax, k = current->fast_outs(kmax); k < kmax; k++) {
+          Node* use = current->fast_out(k);
+          if (use != phi && ctrl_or_self(use) == cl) {
+            DEBUG_ONLY( current->dump(-1); )
+            assert(false, "reduction has use inside loop");
+            break; // Chain traversal fails.
+          }
+        }
+      } else {
+        if (current->outcnt() != 1) {
+          DEBUG_ONLY( current->dump(-1); )
+          assert(false, "reduction (not last) has more than one use");
+          break; // Chain traversal fails.
+        }
+      }
+
+      // Expect another UnorderedReduction or phi as the scalar input.
+      Node* scalar_input = current->in(1);
+      if (scalar_input->is_UnorderedReduction() &&
+          scalar_input->Opcode() == current->Opcode()) {
+        // Move up the UnorderedReduction chain.
+        current = scalar_input->as_UnorderedReduction();
+      } else if (scalar_input == phi) {
+        // Chain terminates at phi.
+        first_ur = current;
+        current = nullptr;
+        break; // Success.
+      } else {
+        DEBUG_ONLY( current->dump(1); )
+        assert(false, "scalar_input is neither phi nor a matchin reduction");
+        break; // Chain traversal fails.
+      }
+    }
+    if (current != nullptr) {
+      // Chain traversal was not successful.
+      continue;
+    }
+    assert(first_ur != nullptr, "must have successfully terminated chain traversal");
+
+    // Determine types
+    const TypeVect* vec_t = last_ur->vect_type();
+    uint vector_length    = vec_t->length();
+    BasicType bt          = vec_t->element_basic_type();
+    const Type* bt_t      = Type::get_const_basic_type(bt);
+
+    // Create vector of identity elements (zero for add, one for mul, etc)
+    Node* identity_scalar = ReductionNode::make_identity_input_for_reduction_from_vector_opc(_igvn, last_ur->Opcode(), bt);
+    set_ctrl(identity_scalar, C->root());
+    VectorNode* identity_vector = VectorNode::scalar2vector(identity_scalar, vector_length, bt_t);
+    _igvn.register_new_node_with_optimizer(identity_vector);
+    set_ctrl(identity_vector, C->root());
+    assert(vec_t == identity_vector->vect_type(), "matching vector type");
+#ifdef ASSERT
+    if (TraceNewVectors) {
+      tty->print("new Vector node: ");
+      identity_vector->dump();
+    }
+#endif
+
+    // Traverse down the chain of UnorderedReductions, and create the vector_accumulators.
+    current = first_ur;
+    Node* last_vector_accumulator = phi;
+    while (true) {
+      // Create vector_accumulator to replace current.
+      Node* vector_input = current->in(2);
+      VectorNode* vector_accumulator = current->make_normal_vector_op(last_vector_accumulator, vector_input, vec_t);
+      _igvn.register_new_node_with_optimizer(vector_accumulator);
+      C->copy_node_notes_to(vector_accumulator, current);
```

Comment:
Can be combined with line 4272 by `_igvn.register_new_node_with_optimizer(vector_accumulator, current);`

Reply:
@pfustc thanks for the hint, I did not know that!